odoo/sql_db.py

```diff
@@ -95,31 +95,63 @@ class BaseCursor:
     """ Base class for cursors that manages pre/post commit/rollback hooks. """
 
     def __init__(self):
-        self.precommit = tools.GroupCalls()
-        self.postcommit = tools.GroupCalls()
-        self.prerollback = tools.GroupCalls()
-        self.postrollback = tools.GroupCalls()
+        self._precommit = tools.GroupCalls()
+        self._postcommit = tools.GroupCalls()
+        self._prerollback = tools.GroupCalls()
+        self._postrollback = tools.GroupCalls()
+
+    def add_precommit(self, func, *types):
+        """ Add a pre-commit hook. """
+        return self._precommit.add(func, *types)
+
+    def add_postcommit(self, func, *types):
+        """ Add a post-commit hook. """
+        return self._postcommit.add(func, *types)
+
+    def add_prerollback(self, func, *types):
+        """ Add a pre-rollback hook. """
+        return self._prerollback.add(func, *types)
+
+    def add_postrollback(self, func, *types):
+        """ Add a post-rollback hook. """
+        return self._postcommit.add(func, *types)
+
+    def precommit(self):
+        """ Process all pre-commit hooks, and discard pre-rollback hooks. """
+        self.flush_env()
```

Comment:
`flush_env` still seems to be only a free function, how can this work?

Reply:
It does not... stupid typo.